src/routing/mod.rs

```diff
@@ -626,24 +626,49 @@ where
     }
 }
 
-#[derive(Debug)]
-pub(crate) struct UrlParams(pub(crate) Vec<(ByteStr, ByteStr)>);
+// we store the potential error here such that users can handle invalid path
+// params using `Result<Path<T>, _>`. That wouldn't be possible if we
+// returned an error immediately when decoding the param
+pub(crate) struct UrlParams(
+    pub(crate) Result<Vec<(ByteStr, PercentDecodedByteStr)>, InvalidUtf8InPathParam>,
```

Comment:
I wonder whether this should be `Vec<Result<(ByteStr, PercentDecodedByteStr), InvalidUtf8InPathParam>>`. Would make sense even without really "supporting" non-UTF8 segments if the path deserializer supported [`IgnoredAny`](https://docs.rs/serde/1.0.130/serde/de/struct.IgnoredAny.html).

Reply:
I think I'll just leave it as it. Can always improve things if someone has a use case.